chain/rust/src/builders/redeemer_builder.rs

```diff
@@ -0,0 +1,330 @@
+    pub fn new() -> Self {
+        Self::default()
+    }
+    
+    pub fn is_empty(&self) -> bool {
+        self.spend.is_empty() && self.mint.is_empty() && self.reward.is_empty() && self.cert.is_empty()
+    }
+
+    /// note: will override existing value if called twice with the same key
+    pub fn update_ex_units(&mut self, key: RedeemerWitnessKey, ex_units: ExUnits) {
+        match key.tag {
+            RedeemerTag::Spend => {
+                let entry = self.spend.iter_mut().nth(key.index as usize).unwrap().1;
+                *entry = Some(UntaggedRedeemerPlaceholder::Full(UntaggedRedeemer::new(entry.as_ref().unwrap().data().clone(), ex_units)));
+            },
+            RedeemerTag::Mint => {
+                let entry = self.mint.iter_mut().nth(key.index as usize).unwrap().1;
+                *entry = Some(UntaggedRedeemerPlaceholder::Full(UntaggedRedeemer::new(entry.as_ref().unwrap().data().clone(), ex_units)));
+            }
+            RedeemerTag::Cert => {
+                let entry = self.cert.iter_mut().nth(key.index as usize).unwrap();
+                *entry = Some(UntaggedRedeemerPlaceholder::Full(UntaggedRedeemer::new(entry.as_ref().unwrap().data().clone(), ex_units)));
+            }
+            RedeemerTag::Reward => {
+                let entry = self.reward.iter_mut().nth(key.index as usize).unwrap().1;
+                *entry = Some(UntaggedRedeemerPlaceholder::Full(UntaggedRedeemer::new(entry.as_ref().unwrap().data().clone(), ex_units)));
+            }
+        }
+    }
+
+    pub fn add_spend(&mut self, result: &InputBuilderResult) {
+        let plutus_data = {
+            result.aggregate_witness.as_ref().and_then(|data| data.plutus_data())
+        };
+        if let Some(data) = plutus_data {
+            self.spend.insert(result.input.clone(), Some(UntaggedRedeemerPlaceholder::JustData(data.clone())));
+        } else {
+            self.spend.insert(result.input.clone(), None);
+        }
+    }
+
+    pub fn add_mint(&mut self, result: &MintBuilderResult) {
+        let plutus_data = {
+            result.aggregate_witness.as_ref().and_then(|data| data.plutus_data())
+        };
+        if let Some(data) = plutus_data {
+            self.mint.insert(result.policy_id.clone(), Some(UntaggedRedeemerPlaceholder::JustData(data.clone())));
+        }
+        else {
+            self.mint.insert(result.policy_id.clone(), None);
+        }
+    }
+
+    pub fn add_reward(&mut self, result: &WithdrawalBuilderResult) {
+        let plutus_data = {
+            result.aggregate_witness.as_ref().and_then(|data| data.plutus_data())
+        };
+        if let Some(data) = plutus_data {
+            self.reward.insert(result.address.clone(), Some(UntaggedRedeemerPlaceholder::JustData(data.clone())));
+        } 
+        else {
+            self.reward.insert(result.address.clone(), None);
+        }
+        
+    }
+
+    pub fn add_cert(&mut self, result: &CertificateBuilderResult) {
```

Comment:
I noticed this (as well as `TransactionWitnessSetBuilder`) are different from the other builders API-wise. @SebastienGllmt do you remember if there was a reason for this? They modify and don't return, whereas other ones return self to allow for function chaining. Was there a performance difference on WASM due to all the cloning instead of moving like in rust? Or some other reason?

Reply:
I don't recall any deep reason. These were originally meant to be Rust-only I think but I haven't double-checked how they will be used going forward